Cython/Compiler/ExprNodes.py

```diff
@@ -6632,16 +6632,68 @@ class PyMethodCallNode(CallNode):
     # arg_tuple   TupleNode     the arguments for the args tuple
     # kwdict      ExprNode or None  keyword dictionary (if present)
     # unpack      bool
+    # function_obj ExprNode or None   copy of self.function.obj (in some situations)    
 
-    subexprs = ['function', 'arg_tuple', 'kwdict']
+    subexprs = ['function', 'arg_tuple', 'kwdict', 'function_obj']
     is_temp = True
     kwdict = None
+    direct_fallback = False
+    function_obj = None
+
+    def __init__(self, pos, **kwds):
+        super().__init__(pos, **kwds)
+        if self.allow_direct_fallback_call():
+            self.direct_fallback = True
+            self.function_obj = ProxyNode(self.function.obj)
+            self.function.obj = CloneNode(self.function_obj)
+
+    def allow_direct_fallback_call(self):
+        if self.kwdict and not isinstance(self.kwdict, DictNode):
+            return False
+        if not self.function.is_attribute:
+            return False
+        return True
+    
+    def generate_direct_fallback_guard(self, code, negate):
+        if self.direct_fallback:
+            # negate gives the guard for the regular code
+            code.putln(f"#if {'!' if negate else ''}(CYTHON_VECTORCALL && !CYTHON_UNPACK_METHODS)")
+
+    def generate_direct_fallback(self,  code, len_args, keyword_variable):
+        if not self.direct_fallback:
+            return
+        code.putln("#else")
+        # The assumption here is that if we can use vectorcall, but can't unpack methods
+        # then PyObject_VectorcallMethod is better because it does most of this stuff internally.
+        # This is especially the case for the Limited API where method creation is expensive
+        code.putln(f"{self.result()} = PyObject_VectorcallMethod("
+                   f"{code.get_py_string_const(self.function.attribute)}, "
+                   f"__pyx_callargs, {len_args+1}{keyword_variable or ', NULL'});")
+        
+    def generate_direct_fallback_self_assignment(self, code, self_arg, arg_offset_cname):
+        if not self.direct_fallback:
+            return
+        self.generate_direct_fallback_guard(code, negate=False)
+        code.putln(f"{self_arg} = {self.function_obj.result()};")
+        code.put_incref(self_arg, py_object_type)
+        code.putln(f"CYTHON_UNUSED_VAR({arg_offset_cname});")
```

Comment:
It seems weird to pass `arg_offset_cname` just to mark it as unused. That creates a very tight dependency on the calling code and the alternative calling code. It's the caller and user of the argument that should mark the argument as unused, not the fallback code that does not use (or should even know about) it.